core/agentbinary/types.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package agentbinary
+
+import (
+	"github.com/juju/version/v2"
+
+	"github.com/juju/juju/core/arch"
+	coreerrors "github.com/juju/juju/core/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// Version represents the version of an agent binary. [Version] was created so
+// that Juju can move itself off of [version.Binary] as this contains the
+// release field that we know longer want.
+type Version struct {
+	// Number is the juju version number.
+	Number version.Number
+
+	// Arch is the architecture of the agent binary.
+	Arch arch.Arch
+}
+
+// Validate checks that the version is valid by asserting the version number is
+// a non zero value and the architecture is supported. If these checks aren't
+// satisfied an error satisfying [coreerrors.NotValid] will be returned.
+func (v Version) Validate() error {
+	if version.Zero == v.Number {
+		return errors.Errorf(
+			"version number cannot be zero value",
+		).Add(coreerrors.NotValid)
+	}
+
+	if !arch.IsSupportedArch(v.Arch) {
+		return errors.Errorf(
+			"unsupported architecture %q",
+			v.Arch,
+		).Add(coreerrors.NotValid)
```

Comment:
.Add(coreerrors.NotSupported)

Reply:
Valiate() doesn't always have to return a not-valid error.
If the validation failure is more specific or contextual, returning a distinct error can make the caller decide what to do with the error.